src/dune_rules/gen_rules.ml

```diff
@@ -386,35 +386,41 @@ type for_melange =
     stanza : Melange_stanzas.Emit.t
   }
 
-let rec nearest_parent_melange_emit ~dir =
+let rec nearest_parent_melange_emits ~dir =
   match Path.Build.parent dir with
-  | None -> Memo.return None
+  | None -> Memo.return []
   | Some parent -> (
     let* stanzas = Only_packages.stanzas_in_dir parent in
     match stanzas with
-    | None -> nearest_parent_melange_emit ~dir:parent
+    | None -> nearest_parent_melange_emits ~dir:parent
     | Some stanzas -> (
       match
-        List.find_map stanzas.stanzas ~f:(function
-          | Melange_stanzas.Emit.T mel -> Some mel
+        List.filter_map stanzas.stanzas ~f:(function
+          | Melange_stanzas.Emit.T stanza ->
+            Some { stanza_dir = parent; stanza }
           | _ -> None)
       with
-      | None -> nearest_parent_melange_emit ~dir:parent
-      | Some stanza -> Memo.return @@ Some { stanza_dir = parent; stanza }))
+      | [] -> nearest_parent_melange_emits ~dir:parent
+      | xs -> Memo.return xs))
 
 (* Detect if [dir] is under the target directory of a melange.emit stanza. *)
 let rec under_melange_emit_target ~dir =
-  let* nearest_parent_melange_emit = nearest_parent_melange_emit ~dir in
-  match nearest_parent_melange_emit with
-  | None -> Memo.return None
-  | Some for_melange ->
-    let target_dir =
-      Melange_rules.emit_target_dir ~dir:for_melange.stanza_dir
-        for_melange.stanza
-    in
-    if Path.Build.is_descendant dir ~of_:target_dir then
-      Memo.return (Some for_melange)
-    else under_melange_emit_target ~dir:for_melange.stanza_dir
+  let* nearest_parent_melange_emits = nearest_parent_melange_emits ~dir in
+  match nearest_parent_melange_emits with
+  | [] -> Memo.return None
+  | for_melange_stanzas -> (
+    match
+      List.find_opt for_melange_stanzas ~f:(fun for_melange ->
+          let target_dir =
+            Melange_rules.emit_target_dir ~dir:for_melange.stanza_dir
+              for_melange.stanza
+          in
+          Path.Build.is_descendant dir ~of_:target_dir)
+    with
+    | Some for_melange -> Memo.return (Some for_melange)
+    | None ->
+      let for_melange_in_this_dir = List.hd for_melange_stanzas in
```

Comment:
The "for_melange" prefixes used across values in this function are a bit confusing to me. What do they mean? Would it make sense to replace them with "emit" or just removed?

Reply:
`for_melange` refers to the `type for_melange` just above the diff (you need to expand it once).